flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java

```diff
@@ -266,10 +271,22 @@ private void runApplicationEntryPoint(
             final Set<JobID> tolerateMissingResult,
             final DispatcherGateway dispatcherGateway,
             final ScheduledExecutor scheduledExecutor,
-            final boolean enforceSingleJobExecution) {
+            final boolean enforceSingleJobExecution,
+            final boolean submitFailedJobOnApplicationError) {
+        if (submitFailedJobOnApplicationError && !enforceSingleJobExecution) {
+            dispatcherGateway.submitFailedJob(
+                    ZERO_JOB_ID,
+                    FAILED_JOB_NAME,
+                    new IllegalStateException(
+                            String.format(
+                                    "Submission of failed job in case of an application error ('%s') is not supported in non-HA setups.",
```

Comment:
The main reason was not having to think about scenarios when the driver can actually submit more than one job.

For example:
- If the exception happens between first and second submission (first one has already completed). What job id do we submit the job with?
- Choosing the submission id is tricky here in general, as we can't really use the `$internal.pipeline.job-id` (which implies the "single job mode")

Any thoughts on this?

Reply:
The intention was actually having a way to communicate this limitation to the user, but I see that using ZERO_JOB_ID is wrong 🤔I think failing the future here might be actually a better option here 👍

Reply:
Completing the `jobIdsFuture` exceptionally here seems right

Reply:
> If the exception happens between first and second submission (first one has already completed). What job id do we submit the job with?

If the job had run the job ID would be random as well, right? Couldn't we use that then?

> using ZERO_JOB_ID might not be correct

We should try to reduce this usage as much as possible, because it is quite problematic (e.g., it breaks archiving).

Reply:
> If the job had run the job ID would be random as well, right? Couldn't we use that then?

For this to be useful, the user should know the jobId upfront (that's one of the reasons for supporting the single execution mode only). Also this is not really an exception from the "application driver", but just an unsupported combination of configurations.

I think the current approach should be sufficient for now (failing the whole dispatcher bootstrap). Also it's an experimental feature, so we can reiterate on this later if we find this confusing.

Reply:
> For this to be useful, the user should know the jobId upfront

I don't think that's really true; outside of application mode users don't know the job ID upfront.
The job name & stacktrace are the identifiable bits imo.

* the stacktrace provides information on where it failed
* the job name could be something like "Job #4" or "Job after <insert job name of previous successful job>", "UserClass#Line<where execute was called>", anything that is reasonable deterministic.

> I think the current approach should be sufficient for now (failing the whole dispatcher bootstrap)

It's not documented though ;) Neither for users (config docs) and devs (comment).